packages/flutter_test/lib/src/controller.dart

```diff
@@ -95,6 +104,104 @@ class SemanticsController {
     return result;
   }
 
+  /// Attemps to find a [SemanticsNode] with the given `action` as close to the
+  /// [SemanticsNode] of `finder` as possible.
+  ///
+  /// The [SemanticsNode] directly related to an [Element] is not necessarily
+  /// the one that can be acted upon. This method finds the directly related
+  /// node and checks whether it supports the given `action`. If the directly
+  /// related node does not support the given `action`, then the closest
+  /// descendent of that node that does support the `action` is returned.
```

Comment:
Why do we need to look at descendants? Couldn't that actually end up targeting the wrong node? Seems dangerous... Maybe we should instead require more specific finders to target the right node?

Reply:
Although, I am not fully understanding the problem yet that searching through the descendent is solving. Could you explain?

Reply:
At a high level, the issue is that some of the interactable `SemanticsNode`s are either un-findable through their `Element` (because the Element is related to a private class), or the `SemanticsNode` doesn't have a related `Element` at all.

For example, for a `ListView` neither `find.byType(ListView)` nor `find.byType(Scrollable)` actually relate to an interactable `SemanticsNode` because the node that actually handles actions is deeper in the hierarchy, possibly related to a private class.

I initially tried to keep it closer to the existing finder structure by using something like `        find.byElementPredicate((element) => element.findRenderObject()?.debugSemantics?.getSemanticsData().hasAction(SemanticsAction.scrollDown) ?? false);` to directly go over the elements in the tree and find a related node, but wasn't able to find the interactable part of a `ListView` that way.

For some reason, it's at least _difficult_ to get the appropriate node without doing some sort of traversal of the Semantics tree directly.

9+ times out of 10, these `findWithAction` methods will return the same node as `find` returns, but this allows for cleaner `Finder`s to be used when you need to interact with things that have a more complex Semantics structure under the hood.

Reply:
I could _try_ to find a consistent limiter to the depth of the search, maybe through the `SemanticsNode.owner` property, but I was concerned that anything I found there would be a heuristic at best and may lead to the same "can't make a finder that relates to the proper `SemanticsNode` issue.

As it currently stands, the `finder` limits where the search _starts_, so it'll never pop back up further into the tree and find completely unrelated things to interact with, but something like a tappable card with a separately tappable control inside it may run into the wrong node issue you mentioned if the outer element doesn't support the `tap` action at the time of the call.

Reply:
I am a little concerned about always implicitly walking the descendants to perform a semantics action. This seems like it could be an easy source to miss bugs: e.g. my unit test "verifies" that I can perform a semantics tap action on a thing labeled "click me", when in reality a a11y user focused on the semantics node with the "click me" text can't actually perform the tap because due to a bug, the tap action ended up on a child node.

I think in the general case we should be encouraging people to write semantics action tests using semantics-related finders that target the node with the action directly, examples:

```dart
// Makes sure that if you're focused on a semantics node that reads out "click me" as label
// you can perform a semantics tap on that node.
tester.semantics.tap(find.bySemanticsLabel('Click me'));

// Makes sure that if you're focused on a semantics node that reads out "click me" as value
// you can perform a semantics tap on that node.
tester.semantics.tap(find.bySemanticsValue(10));

// Makes sure that if you're focused on a semantics node that identifies as TextField
// you can perform a semantics tap on that node.
tester.semantics.tap(find.hasSemanticsFlag(SemanticsFlag.isTextField));
```

I believe this should work without walking any descendants for most actions and SemanticsNodes - even if their Element class is private. It doesn't work for SemanticsNodes that don't have a corresponding Element, though, you're right.

Scrolling actions should probably be modeled similarly. With most a11y services you can scroll when something inside of the scrollable has focus, i.e. in a test for a list of tiles 1..10 I would like to write something like:

```dart
tester.semantics.scrollUpSurroundingScrollable(find.bySemanticsLabel('Tile 1'));
```

In other words, I think for scrolling the search should go the other way around to model real a11y services behavior. You find something INSIDE the scrollable and then say, "Hey, scroll the surrounding scrollable of this thing up". I believe, the current implementation assumes it is the other way around: you find something above the scrollable to scroll it.

In summary, I think the tests should be written from the perspective of an a11y user: You find something that the a11y would focus using semantics-related finders and then check that the correct actions can be triggered on that particular node.

Now, this still leaves us with the problem of SemanticsNodes that are not owned by an Element because those cannot be targeted with a finder, as you rightfully state. Those are tough. It's a somewhat rare case, though. Quick survey of the framework shows that we only do this for Scrollables, RenderParagraph, RenderEditable, CustomPainters, and RangeSlider. I wonder, if for these rare cases it would be ok to just have test authors walk the semantics tree manually, but it should be explicit. E.g. you'd get the enclosing SemanticsNode via an Element-finder and then manually walk the children. We could provide a helper for that (your `findWithAction` seems helpful for that use case, maybe renamed to `findDescendantWithAction`). We could also provide a more general `findDescendantWithPredicate`. I imagine, that could be useful for the RenderParagraph case where you want to tap a link, something like:

```dart
tester.semantics.performAction(
  SemanticsAction.tap, 
  tester.semantics.findDescendantWithPredicate(find.type(Text), (node) => node.label == 'link text'),
);
```

In the general use case you'd target the semantics node defining the action directly with a semantics-related finder. In the (hopefully rare) special cases, you'd have to do a bit of tree walking with a provided helper method.

Reply:
@goderbauer and I went over this thread on GVC to figure out a path forward. This is the summary of what we're planning next:

* Remove the implicit downward traversal for all action methods except the scrolling actions
* Update the scrolling actions to be similar to `scrollUntilVisible`, to ensure the semantics API is similar to the non-semantics API and feels familiar to developers
  * Remove the current `finder` parameter, replacing it with an optional `scrollable` parameter
  * If `scrollable` is not provided, then we will search for the first `Scrollable` in the element tree
  * Once `scrollable` is evaluated, traverse down the semantics tree to the nearest descendant that can scroll and use that to execute the SemanticsAction
* Instead of `findWithAction` (or possibly in addition to), we will expose a `findDescendantByPredicate` method, which will be used to traverse the semantics tree (instead of the element tree) both internally and by users

Reply:
While attempting to implement this plan, I found that the current `Finder` based API is just not consistent enough to find specific `SemanticsNode`s. After discussion with @goderbauer I've filed #123634 to introduce a `SemanticsFinder` API that would be better suited to this work.

Resolving this and other discussions for now and putting this PR on the back-burner until that API is available.